Subject: [PATCH] git-instaweb: add git-browser graphiclog link

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 git-instaweb.sh | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/git-instaweb.sh b/git-instaweb.sh
index 5ac32127..f771a321 100755
--- a/git-instaweb.sh
+++ b/git-instaweb.sh
@@ -106,6 +106,7 @@ start_httpd () {
 
 	# generate correct config file if it doesn't exist
 	test -f "$conf" || configure_httpd
+	test -f "$fqgitdir/gitweb/git-browser.conf" || gitbrowser_conf
 	test -f "$fqgitdir/gitweb/gitweb_config.perl" || gitweb_conf
 
 	# don't quote $full_httpd, there can be arguments to it (-f)
@@ -197,7 +198,8 @@ mkdir -p "$GIT_DIR/gitweb/tmp"
 GIT_EXEC_PATH="$(git --exec-path)"
 GIT_DIR="$fqgitdir"
 GITWEB_CONFIG="$fqgitdir/gitweb/gitweb_config.perl"
-export GIT_EXEC_PATH GIT_DIR GITWEB_CONFIG
+GITBROWSER_CONFIG="$fqgitdir/gitweb/git-browser.conf"
+export GIT_EXEC_PATH GIT_DIR GITWEB_CONFIG GITBROWSER_CONFIG
 
 webrick_conf () {
 	# webrick seems to have no way of passing arbitrary environment
@@ -263,7 +265,7 @@ server.force-lowercase-filenames = "disable"
 # variable above and uncomment this
 #accesslog.filename = "$fqgitdir/gitweb/$httpd_only/access.log"
 
-setenv.add-environment = ( "PATH" => env.PATH, "GITWEB_CONFIG" => env.GITWEB_CONFIG )
+setenv.add-environment = ( "PATH" => env.PATH, "GITWEB_CONFIG" => env.GITWEB_CONFIG, "GITBROWSER_CONFIG" => env.GITBROWSER_CONFIG )
 
 cgi.assign = ( ".cgi" => "" )
 
@@ -592,6 +594,13 @@ EOF
 	rm -f "$conf"
 }
 
+gitbrowser_conf() {
+	cat > "$fqgitdir/gitweb/git-browser.conf" <<EOF
+gitbin: $(command -v git)
+warehouse: $(dirname "$fqgitdir")
+EOF
+}
+
 gitweb_conf() {
 	cat > "$fqgitdir/gitweb/gitweb_config.perl" <<EOF
 #!@@PERL@@
@@ -607,6 +616,9 @@ our \$favicon = "/static/git-favicon.png";
 our \$javascript = "/static/gitweb.js";
 our @diff_opts = ('-B', '-C');
 
+\$feature{'actions'}{'default'}=[
+	('graphiclog', "/git-browser/by-commit.html?r=%n", 'log')
+];
 \$feature{'remote_heads'}{'default'} = [1];
 \$feature{'blame'}{'default'} = [1];
 \$feature{'pathinfo'}{'default'} = [1];
@@ -668,6 +680,7 @@ restart)
 	;;
 esac
 
+gitbrowser_conf
 gitweb_conf
 
 resolve_full_httpd
-- 
tg: (f68f1c5f..) t/instaweb/git-browser (depends on: t/instaweb/defaults)
