Subject: [PATCH] git-instaweb: allow running in subdir

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 git-instaweb.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/git-instaweb.sh b/git-instaweb.sh
index 513efa66..4c0af04f 100755
--- a/git-instaweb.sh
+++ b/git-instaweb.sh
@@ -20,6 +20,7 @@ start          start the web server
 restart        restart the web server
 "
 
+SUBDIRECTORY_OK=Yes
 . git-sh-setup
 
 fqgitdir="$GIT_DIR"
-- 
tg: (a26d01e1..) t/instaweb/subdir (depends on: release)
