diff --git a/tg.sh b/tg.sh
index 471e9a83..3a0f1584 100644
--- a/tg.sh
+++ b/tg.sh
@@ -3,7 +3,7 @@
 # (c) Petr Baudis <pasky@suse.cz>  2008
 # GPLv2
 
-TG_VERSION=0.10
+TG_VERSION=0.11
 
 # Update if you add any code that requires a newer version of git
 GIT_MINIMUM_VERSION=1.7.7.2
@@ -37,6 +37,11 @@ precheck() {
 	    || die "git version >= ${GIT_MINIMUM_VERSION} required"
 }
 
+case "$1" in version|--version|-V)
+	echo "TopGit version $TG_VERSION"
+	exit 0
+esac
+
 precheck
 [ "$1" = "precheck" ] && exit 0
 
@@ -428,13 +433,16 @@ do_help()
 			sep="|"
 		done
 
-		echo "TopGit v$TG_VERSION - A different patch queue manager"
+		echo "TopGit version $TG_VERSION - A different patch queue manager"
 		echo "Usage: tg ( help [<command>] | [-r <remote>] ($cmds) ...)"
-	elif [ -r "@cmddir@"/tg-$1 ] ; then
+		echo "Use \"tg help tg\" for overview of TopGit"
+	elif [ -r "@cmddir@"/tg-$1 -o -r "@sharedir@/tg-$1.txt" ] ; then
 		setup_pager
 		{
-			"@cmddir@"/tg-$1 -h 2>&1 || :
-			echo
+			if [ -r "@cmddir@"/tg-$1 ] ; then
+				"@cmddir@"/tg-$1 -h 2>&1 || :
+				echo
+			fi
 			if [ -r "@sharedir@/tg-$1.txt" ] ; then
 				cat "@sharedir@/tg-$1.txt"
 			fi
