Subject: [PATCH] gitweb: enable case-insensitive sorting

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>
---
 gitweb/gitweb.perl | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gitweb/gitweb.perl b/gitweb/gitweb.perl
index 174a3a28..f79414e9 100755
--- a/gitweb/gitweb.perl
+++ b/gitweb/gitweb.perl
@@ -5950,7 +5950,7 @@ sub sort_projects_list {
 
 	sub order_str {
 		my $key = shift;
-		return sub { $a->{$key} cmp $b->{$key} };
+		return sub { lc($a->{$key}) cmp lc($b->{$key}) };
 	}
 
 	sub order_reverse_num_then_undef {
---
